### PR TITLE
CRM task n-level hierarchies

### PR DESCRIPTION
**Pull request by andersdd**
builds on CRM task migration PR ( xtuple/xtuple#3273) so review and merge that one first

### CODE REVIEW

File: foundation-database/public/tables/metasql/tasklist-detail.mql [hunk 1]
```diff
@@ -0,0 +1,537 @@
...
+           (SELECT array_to_string(array_agg(taskass_username),',') AS assigned FROM taskass WHERE taskass_task_id=task_id) AS assigned,
+           firstLine(task_descrip) AS notes, 
+           task_start_date AS start, 
+           task_due_date AS due, 
+           formatDate(task_due_date) AS f_due, 
+           task_start_date is not null AS hasStartDate,
+           task_due_date is not null AS hasDueDate,
+           NULL AS crmacct_id,
+           '' AS crmacct_number, 
+           '' AS crmacct_name,
```
Comment (gilmoskowitz) [inline foundation-database/public/tables/metasql/tasklist-detail.mql]: Consider moving all of the `LEFT OUTER JOIN`s with `crmacct` and `custinfo` to the outermost select. That would simplify the query execution plan for most parts of the `UNION` and remove some of the maintenance clutter.

File: foundation-database/public/tables/report/ProjectTaskList.xml [hunk 2]
```diff
@@ -17,17 +20,35 @@
...
+       END AS status,
+       formatDate(task_due_date) AS f_due,
+       formatDate((SELECT MIN(taskass_assigned_date) FROM taskass WHERE taskass_task_id=t.task_id)) AS f_assigned,
+       formatDate(task_start_date) AS f_start,
+       formatDate(task_completed_date) AS f_completed,
+       path_info
+  FROM task_hier h
+  JOIN task t ON t.task_id=h.task_id
+  WHERE (task_parent_id = &lt;?value("prj_id")?> AND task_parent_type='J')
+  ORDER BY path_info</sql>
```
Comment (gilmoskowitz) [inline foundation-database/public/tables/report/ProjectTaskList.xml]: Can this reuse `orderActivityByProject-tasks.mql`?

File: foundation-database/public/trigger_functions/prospect.sql [hunk 3]
```diff
@@ -132,18 +132,6 @@ BEGIN
...
-                            (SELECT opstage_name FROM opstage WHERE opstage_id=OLD.prospect_opstage_id),
-                            (SELECT opstage_name FROM opstage WHERE opstage_id=NEW.prospect_opstage_id));
-      END IF;
-
-      IF (OLD.prospect_priority_id <> NEW.prospect_priority_id) THEN
-        PERFORM postComment('ChangeLog', 'PSPCT', NEW.prospect_id, 'Priority',
-                            (SELECT incdtpriority_name FROM incdtpriority WHERE incdtpriority_id=OLD.prospect_priority_id),
-                            (SELECT incdtpriority_name FROM incdtpriority WHERE incdtpriority_id=NEW.prospect_priority_id));
-      END IF;
-
```
Comment (gilmoskowitz) [inline foundation-database/public/trigger_functions/prospect.sql]: Removing these doesn't feel right. Perhaps a comment is posted elsewhere for these changes.

File: foundation-database/public/trigger_functions/task.sql [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+                                      FROM taskass 
+                                     WHERE taskass_task_id=NEW.task_id)), false)) THEN
+      RAISE EXCEPTION 'You do not have privileges to maintain Tasks.';
+    END IF;
+  ELSE  
+    RAISE EXCEPTION 'You do not have privileges to maintain Tasks.';
+  END IF;
+
+  IF (LENGTH(COALESCE(NEW.task_number,'')) = 0) THEN
+    RAISE EXCEPTION 'You must enter a valid number.';
```
Comment (gilmoskowitz) [inline foundation-database/public/trigger_functions/task.sql]: Consider checking if task numbers are configured for automatic or auto-with-override. If so, just generate one here.

File: foundation-database/public/functions/createtask.sql [hunk 5]
```diff
@@ -0,0 +1,134 @@
+DROP FUNCTION IF EXISTS createtask(TEXT, TEXT, TEXT, TEXT, BOOLEAN, TEXT, TEXT, TEXT, TEXT,
+                                 JSON, NUMERIC, NUMERIC, NUMERIC, NUMERIC, NUMERIC, DATE, DATE,
+                                 DATE, TEXT) CASCADE;
+DROP FUNCTION IF EXISTS createtask(TEXT, TEXT, TEXT, TEXT, TEXT, TEXT, TEXT, TEXT,
+                                 JSON, NUMERIC, NUMERIC, NUMERIC, NUMERIC, NUMERIC, DATE, DATE,
+                                 DATE, TEXT) CASCADE;
+
+CREATE OR REPLACE FUNCTION createtask (
+    pParentType TEXT,
+    pParent TEXT,
```
[the remark this answers is not in the document]
Comment (andersdd) [reply foundation-database/public/functions/createtask.sql]: `createtask` is used by the API view which receives human-readable numbers not internal IDs.  In this case it receives the `document_number` not the `document_id`.  Whilst most CRM document numbers _are_ integers, not all of them are (project task numbers are alphanumeric)

File: foundation-database/public/functions/createtaskfromopportunity.sql [hunk 6]
```diff
@@ -0,0 +1,25 @@
+CREATE OR REPLACE FUNCTION createTaskFromOpportunity(pOpheadId INTEGER, pTaskName TEXT)
+RETURNS INTEGER AS $$
+-- Copyright (c) 1999-2018 by OpenMFG LLC, d/b/a xTuple. 
+-- See www.xtuple.com/CPAL for the full text of the software license.
+DECLARE
+  taskid   INTEGER;
+BEGIN
+  taskid := (SELECT createtask('OPP'::TEXT, pOpheadId::TEXT, pTaskName, 'N'::TEXT, true, pTaskName, pTaskName,
```
[the remark this answers is not in the document]
Comment (andersdd) [reply foundation-database/public/functions/createtaskfromopportunity.sql]: Not sure what is going on here.  I removed the task `active` field as we also have status.  The `createtaskfromopportunity()` function in my development system correctly calls the revised `createtask()` function.  The source code is correct, but not reflected in Github (all the code has been pushed though).

File: foundation-database/public/trigger_functions/prospect.sql (hunk 3 above)
Comment (andersdd) [reply foundation-database/public/trigger_functions/prospect.sql]: Priority was originally added as part of CRM Enhancements but we removed from Prospect as the priority of work is indicated on a Prospect's Opportunity

File: foundation-database/public/tables/metasql/orderActivityByProject-orders.mql [hunk 7]
```diff
@@ -0,0 +1,1214 @@
...
+<? if exists("owner_username") ?>
+    JOIN prj ON (prj_id=poitem_prj_id
+            AND ((prj_owner_username=<? value("owner_username") ?>) OR (prj_username=<? value("owner_username") ?>)))
+<? endif ?>
+ WHERE (poitem_prj_id = <? value("prj_id") ?>)
+
+<? endif ?>
+
+) data
+ORDER BY section, subtype, type, id;
```
[the remark this answers is not in the document]
Comment (andersdd) [reply foundation-database/public/tables/metasql/orderActivityByProject-orders.mql]: This is a stripped down version of `OrderActivityByProject-detail` which is the single biggest/worst metaSQL in the system.  Do subtotals work in XTreeWidget.  We have a couple of fields indicating level/type that might be useful, although this nested query approach allows for specific text to be created and allows specific grouping of sales orders/invoices under an Orders subtotal.

**Comment by gilmoskowitz** [comment]
OK to merge after merge conflicts have been resolved.

**Pull request merged**